Add temp humidity sensors

kylerove opened the pull request:
Add separate humidity and temperature sensors from thermostats for both inside and outside (see https://github.com/NorthernMan54/homebridge-tcc/issues/76). Because outside values are repeated for each thermostat, needed to add config options for each thermostatID to allow users to only show sensors they are interested in. This took some time to get the code right and I was fighting a lot of the config conditional code and where to put it.

One issue I worked around was in `TccAccessory()` if the devices already exist in the Homebridge instance, we still needed to check for any changes in the config code to add sensors to reflect changes. I have not tested what it means if a user changes config for a sensor from true to false. Probably need a way to detect this and delete the cached accessory/service.

Otherwise, this works great!

NorthernMan54 commented:
Tks for submitting this.  I have a couple of comments before merging this

1 - I understand your logic around having the deviceID in the config, but I'm not a fan of having people look at logs as part of the configuration of an accessory.  Is there an approach to include them without requiring the deviceID ?  Maybe a basic and advanced approach ?

2 - What about multiple indoor sensors ?

kylerove commented:
Yeah not my favorite either. We could create a top level setting that turns on all indoor, all outdoor sensors, all sensors or none (default). The config I proposed could be additional advanced setting to allow for granular control.

kylerove commented:
OK, config is simpler. Updated Read Me to reflect sensor options and advanced config options for those who want granular control. The cached accessory thing in Homebridge can be difficult to work around when developing! Couldn't figure out why I had duplicate outdoor sensors despite new code to limit to 1 set. Finally removed the cached accessories and that did it.

dbavatar commented:
Trying this out. It added my outdoor temp/humidity to the first enumerated thermostat, but now that means that Home thinks that's the temperature in that room. Perhaps it would be better to add them as standalone sensors that are not attached to a thermostat?

kylerove commented:
This behavior is sort of how Homebridge works. All accessories get dumped into the room the bridge is assigned. For example, I have 3 thermostats and by default they all showed up in the Basement (where I assigned the homebridge). From the Home app, though, you can move accessories around, including these sensors. Not sure Homebridge has a way to designate which rooms accessories should belong.

dbavatar commented:
Perhaps I'm doing something wrong, but the sensors are attached as accessories to the thermostat, they cannot be moved independently. For example, if I go to the Humidity sensor attached to "Living Room" change move it to "Outside", then all 3 accessories move:

![image](https://user-images.githubusercontent.com/14057304/103489667-0e232a80-4de4-11eb-8076-7b31e403d532.png)

kylerove commented:
My apologies! You are correct. I assumed the room could be independently set for each accessory even when paired. This is rather annoying.

To create sensors that are separate (perhaps just the outside pair?), this will require refactor with copy of `TccAccessory()` function just for said sensors. Perhaps something like `TccSensor()`. Should all the sensors be separate or just the outside pair? I personally think the indoor sensors can continue to be paired, but the outdoor should be a separate accessory to allow for independent room assignment. Thoughts?

dbavatar commented:
I don't have any standalone indoor sensors and my thermostats don't have humidity sensors, so I can't check.

However from what I gather you would want both behaviors configurable, as an add-on sensor that can be placed anywhere like C7189R1004, and you can also have each thermostat expose sensors. In my case my thermostats always report 128 for humidity since the hw doesn't exist, but if it did it would make sense to attach the sensor to the thermostat, like it's doing now.

NorthernMan54 commented:
Technically Service's of a single accessory are grouped into the same room.   So to have the sensors able to be placed into different rooms, they need to be separate accessories in the code.

NorthernMan54 commented:
@kylerove @dbavatar Should I publish this pull request in its current form ?

kylerove commented:
I might have time in the next few days to separate out the outdoor which I think will get it into the best place if you want to hold off. I don't think it will be terribly hard.

kylerove commented:
I refactored the code:
- new `tccSensorsAccessory()`
- because both thermostat and sensors device to update when `pollThermostats()` fires and both include ThermostatID to link the two for the thermostat off which we are getting outside sensor data, `TccAccessory()` and `TccSensorsAccessory()` use `getAccessoryByName()` to check if the accessory/services already exist to create/update.
- removed `advanced` option from config (seems like existing `none`, `all`, `inside`, `outside` will provide enough configurability without introducing too much confusion)
- updated the README

Can now move outside sensors into another room (Yard, Outside, etc) separate from the thermostat from which it gets it's data! Did my best to keep code structure clean and readable. Let me know if we need any other changes!

dbavatar commented:
Tried it out, appears to be working, thanks!

dbavatar commented:
There's a bug, after restarting homebridge, there's a problem with cachedAccessories. Deleting the file lets it start, but it reproduces the problem on the subsequent restart:

`[1/11/2021, 11:15:02 PM] [TCC] configureAccessory Living Room
Tue, 12 Jan 2021 04:15:02 GMT tcc FakeGatoHistoryService fs 60
[1/11/2021, 11:15:02 PM] [TCC] configureAccessory Outside Sensors
[1/11/2021, 11:15:02 PM] [TCC] configureAccessory Bedroom 2
Tue, 12 Jan 2021 04:15:02 GMT tcc FakeGatoHistoryService fs 60
[1/11/2021, 11:15:02 PM] [TCC] configureAccessory Master Bedroom
Tue, 12 Jan 2021 04:15:02 GMT tcc FakeGatoHistoryService fs 60
[1/11/2021, 11:15:02 PM] [TCC] configureAccessory Upper Level
Tue, 12 Jan 2021 04:15:02 GMT tcc FakeGatoHistoryService fs 60
[1/11/2021, 11:15:02 PM] [TCC] configureAccessory Garage
Tue, 12 Jan 2021 04:15:02 GMT tcc FakeGatoHistoryService fs 60
[1/11/2021, 11:15:02 PM] [TCC] configureAccessory Living Room
Tue, 12 Jan 2021 04:15:02 GMT tcc FakeGatoHistoryService fs 60
[1/11/2021, 11:15:02 PM] Error: Cannot add a bridged Accessory with the same UUID as another bridged Accessory: 58204295-2c9f-49fe-9202-db6253105f77
    at Bridge.Accessory._this.addBridgedAccessory (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:419:15)
    at /usr/local/lib/node_modules/homebridge/src/server.ts:308:19
    at Array.filter (<anonymous>)
    at Server.restoreCachedPlatformAccessories (/usr/local/lib/node_modules/homebridge/src/server.ts:268:69)
    at Server.start (/usr/local/lib/node_modules/homebridge/src/server.ts:156:10)
    at cli (/usr/local/lib/node_modules/homebridge/src/cli.ts:80:10)
    at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
[1/11/2021, 11:15:02 PM] Got SIGTERM, shutting down Homebridge...
`

kylerove commented:
Try it now. I found the primary error that was leading to this in the new `getAccessoryByName(accessoryName)` function as I had forgotten to add the device name to the thermostats to search when checking to see if the accessory was cached or not. Also fixed other errors that were being thrown because of the fakegato code. I don't use the Eve app, but I think I was able to right the error with additional conditional checks, even adding logging for the outdoor sensors to mimic Eve Weather.

dbavatar commented:
Seems to be working. Possibly interesting log lines:

1st start:
```
Sat, 16 Jan 2021 16:21:46 GMT tcc-lib TCC - Login Succeeded
Sat, 16 Jan 2021 16:21:47 GMT tcc Creating accessory for Living Room(518496)
Sat, 16 Jan 2021 16:21:47 GMT tcc createInsideSensors:  false
[1/16/2021, 11:21:47 AM] [TCC] Adding TCC Device (deviceID=518496) Living Room
[1/16/2021, 11:21:47 AM] [TCC] Adding TCC Outside Sensors (deviceID=518496) Outside Sensors
Sat, 16 Jan 2021 16:21:47 GMT tcc TccSensorsAccessory() Outside Sensors outsideTemperature = true, existing sensor
Sat, 16 Jan 2021 16:21:47 GMT tcc TccSensorsAccessory() Outside Sensors outsideHumidity = true, existing sensor
Sat, 16 Jan 2021 16:21:47 GMT tcc Creating accessory for Bedroom 2(518597)
Sat, 16 Jan 2021 16:21:47 GMT tcc createInsideSensors:  false
[1/16/2021, 11:21:47 AM] [TCC] Adding TCC Device (deviceID=518597) Bedroom 2
Sat, 16 Jan 2021 16:21:47 GMT tcc Creating accessory for Master Bedroom(520413)
Sat, 16 Jan 2021 16:21:47 GMT tcc createInsideSensors:  false
[1/16/2021, 11:21:47 AM] [TCC] Adding TCC Device (deviceID=520413) Master Bedroom
Sat, 16 Jan 2021 16:21:47 GMT tcc Creating accessory for Upper Level(552358)
Sat, 16 Jan 2021 16:21:47 GMT tcc createInsideSensors:  false
[1/16/2021, 11:21:47 AM] [TCC] Adding TCC Device (deviceID=552358) Upper Level
Sat, 16 Jan 2021 16:21:47 GMT tcc Creating accessory for Garage(1256152)
Sat, 16 Jan 2021 16:21:47 GMT tcc createInsideSensors:  false
[1/16/2021, 11:21:47 AM] [TCC] Adding TCC Device (deviceID=1256152) Garage
HAP Warning: Characteristic 000000AC-0000-1000-8000-0026BB765291 not in required or optional characteristics for service 000000BB-0000-1000-8000-0026BB765291. Adding anyway.
```
subsequent starts:
```
[1/16/2021, 11:30:37 AM] [TCC] Existing TCC accessory (deviceID=518496) Living Room
[1/16/2021, 11:30:37 AM] [TCC] Existing TCC outside sensors accessory (deviceID=518496) Outside Sensors
[1/16/2021, 11:26:08 AM] [TCC] Critical Error - No devices created, please restart.
[1/16/2021, 11:26:08 AM] [TCC] Error: TooManyAttempt
```
```[1/16/2021, 11:30:37 AM] [TCC] Existing TCC accessory (deviceID=518496) Living Room
[1/16/2021, 11:30:37 AM] [TCC] Existing TCC outside sensors accessory (deviceID=518496) Outside Sensors
[1/16/2021, 11:30:37 AM] [TCC] Critical Error - No devices created, please restart.
[1/16/2021, 11:30:37 AM] [TCC] Cannot read property 'addEntry' of undefined```

kylerove commented on index.js:
```diff
@@ -383,7 +372,7 @@ function TccAccessory(that, device, sensors) {
     
     // need to check if accessory/zone/thermostat already exists, but user added temp/humidity sensors then must declare
-    this.accessory = getAccessoryByThermostatID(this.ThermostatID);
+    this.accessory = getAccessoryByName(this.name);
```
Cause of duplicate sensors

kylerove commented:
When I thought this would be straight forward, I didn't account for some of the fakegato code or some of the complexities of mapping n physical thermostats to n+1 accessories. I think I have it.

**As before, you should delete any TCC-created cached accessories.** I tested through several restarts both creating new accessories and using cached accessories. No more duplicate sensors. No errors. I am still getting the `Characteristic undefined...` warning, but it doesn't interfere with the rest of the plugin and the `addEntry` error is gone.

Appreciate your ability to test this and work through the issues and improve the code!

dbavatar commented:
> Appreciate your ability to test this and work through the issues and improve the code!

Thank you for working on it!

Everything seems to work now. Although I think I found a cosmetic issue only visible in the Catalina macOS Home app, I assume there's nothing to do about it. Look good in Big Sur/iOS 14.

![image](https://user-images.githubusercontent.com/14057304/104820504-06f81580-5803-11eb-89c9-7970082f50a0.png)

NorthernMan54 commented:
Are we good ? Should I merge and publish ?

kylerove commented:
I think it's good to go. I've been using since last set of changes without any issues or errors.

dbavatar commented:
LGTM

NorthernMan54 commented:
Tks everyone, publishing
